typescript/src/entities/product.test.ts

```diff
@@ -84,3 +119,35 @@ test('can list products from server with category', () => {
     expect((d[0].categories as any)[0].name).toBe('c1');
   });
 });
+
+test('can save product', () => {
+  const merchi = new Merchi();
+  const c1 = new merchi.Category();
+  const p = new merchi.Product();
+  const c2 = new merchi.Category();
+  const d = new merchi.Domain();
+  p.categories = [c2];
+  p.domain = d;
+  c1.products = [p];
+  c1.save();
+  d.domain = "3onrb6o4";
+  p.name = "pHyz7ZucK#";
+  c2.name = "8&OaUsDgJ$ev3FYZ3";
+  p.save();
+  (global as any).fetch = jest.fn().mockImplementation((url, data) => {
+    expect(data.method).toBe("PATCH");
+    const correct =  [['products-0-name', 'pHyz7ZucK#'],
+      ['products-0-categories-0-name', '8&OaUsDgJ$ev3FYZ3'],
+      ['products-0-categories-count', '1'],
+      ['products-0-domain-0-domain', '3onrb6o4'],
+      ['products-0-domain-count', '1'],
+      ['products-count', '1']];
+    expect(Array.from(data.body.entries())).toEqual(correct);
```

Comment:
it is magic here to me, how did you guarantee Array.from(data.body.entries()) and "correct" has the same order

Another good idea actually bring by @Amtron that we should use a nested json as correct and serialise it instead directly write it as flat out json (hope I explained properly)

Whether the serialization function is correct is a independent thing that needs to test.

this makes the code more readable and scalable, if the protocol changed we do not need to fix all testcases only need to adjust the testcase for serialization function

Reply:
serialization function refers the toFormData you done blow

Reply:
> how did you guarantee Array.from(data.body.entries()) and "correct" has the same order

the order is of course not important to the protocol currently (though there is actually movement to change that in some cases).

`toFormData` however is detirministic, and therefore always traverses the graph in the same order, and therefore always outputs in the same order. MDN doesn't mention anything about the ordering of `FormData.entries()`, but it was be surprising to me if they used a non-deterministic iterator.

Reply:
> that we should use a nested json as correct and serialise it instead directly write it as flat out json 

what's the supposed benefit of splitting the transformation over two steps?

Ultimately these entity objects are just javascript objects the same as JSON obejcts would be. They just contain some extra information, too. Why spend an extra step removing that extra information, when the final step would presumably be the same regardless?